bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/InMemoryRecoveryLog.java

```diff
@@ -0,0 +1,77 @@
+package io.ballerina.runtime.transactions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static io.ballerina.runtime.transactions.TransactionConstants.IN_MEMORY_CHECKPOINT_INTERVAL;
+
+public class InMemoryRecoveryLog implements RecoveryLog {
+     private static final Logger log = LoggerFactory.getLogger(InMemoryRecoveryLog.class);
+    private Map<String, TransactionLogRecord> transactionLogs;
+    private int numOfPutsSinceLastCheckpoint;
+
+    public InMemoryRecoveryLog() {
+        this.transactionLogs = new ConcurrentHashMap<>();
+        this.numOfPutsSinceLastCheckpoint = 0;
+    }
+
+    @Override
+    public void put(TransactionLogRecord trxRecord) {
+        transactionLogs.put(trxRecord.getCombinedId(), trxRecord);
+        ifNeedWriteCheckpoint();
+    }
+
+    /**
+     * Write a checkpoint to the in-memory log (not needed if you don't need checkpoints).
+     */
+    public void ifNeedWriteCheckpoint() {
```

Comment:
Can't we make the `RecoveryLog` an abstract class  and have common methods for `FileRecoveryLog` and `InMemoryRecoveryLog`  defined there.